src/satosa/frontends/saml2.py

```diff
@@ -299,18 +300,36 @@ def _handle_authn_response(self, context, internal_response, idp):
             attributes_to_remove = custom_release.get("exclude", [])
             for k in attributes_to_remove:
                 ava.pop(k, None)
-
-        name_id = NameID(text=internal_response.user_id,
-                         format=hash_type_to_saml_name_id_format(
-                             internal_response.user_id_hash_type),
-                         sp_name_qualifier=None,
-                         name_qualifier=None)
+        idp_conf = self.idp_config.get('service', {}).get('idp', {})
+        try:
+            config_nameidformat = idp_conf['name_id_format']
+            if isinstance(config_nameidformat, list):
+                config_nameidformat = config_nameidformat[0]
+        except KeyError:
+            config_nameidformat = None
+            satosa_logging(logger, logging.DEBUG, "No nameid format configured for frontend idp",
+                           context.state)
+        if config_nameidformat == NAMEID_FORMAT_EMAILADDRESS:
+            nameidfmt = NAMEID_FORMAT_EMAILADDRESS
+            nameidval = internal_response.attributes.get('mail', None)[0]
+            name_id = NameID(text=nameidval,
```

Comment:
`pysaml2` is a different project (although really close to SATOSA). Exception classes defined by pysaml, should only be used by pysaml. SATOSA defines its own classes to handle exception errors and those should be used. 

If there is no matching class that describes the error well, then a new class should be defined, coupled with an appropriate error message. The class describes the family of errors, while the message describes the error. 

In this case the error stems from a misconfiguration of the provider, not SATOSA. For SATOSA this is a SAML module error (saml frontend module cannot process the response), or a requirement error (emailaddress nameid format expects the mail attribute to be set). I see no such error class defined, so you can go forward and define one. 

The exception should be raised and be chained with the underlying cause
```python
except KeyError as e:
  log(..)
  raise FamilyError("specific and descriptive error message") from e
```

---

However, there is another problem I see here. You have hardcoded the 'mail' internal attribute name. Internal attribute names are defined in the configuration and 'mail' may not be there. Internal attributes should not be hardcoded as the user specifies them -  one may prefer to name the attribute 'eMail' for example.. They are not related to the actual attribute names, they only map their values. 

What you want to do is set the `user_id_from_attrs` configuration option in `internal_attributes.yaml` to the list of internal attributes that you want to be used as the _NameID_ value (namely `user_id_from_attrs: [mail]`). Then, internally `internal_response.user_id` will match those attributes values (concatenated). 

---

So, I think this should be left unchanged. The configuration should make sure that the user_id will be generated by the _internal attribute_ that maps to the 'mail' _saml attribute_. Support for the nameid-format `emailAddress` should be added to [UserIdHashType](https://github.com/SUNET/SATOSA/blob/11edb70ef16f70022463b869c9c2219ae9100026/src/satosa/internal_data.py#L10) class.

Reply:
I agree, extending UserIdHashType is better.

I think that I was misled by the name 'UserIdHashType'. I understand that the class tries to generalize the NameID concept across other protocols. However, is it really limited to hashed representations as implied by the name? E.g. SAML does not prescribe to generate a persistent NameID using a hash, you could use AES or databse keys. Also, does the class intend to provide a fuzzy mapping, or should this allow for precise mapping?

Assuming that the class is not limited to hashed UserIDs it should be renamed to UserIdScopedType.

Extending the UserIdHashType enumeration with "emailAddress", together with hash_type_to_saml_name_id_format will do the job. What remains open is that this configuration option is disjointed from the frontend's idp_config/../name_form. Isname_form there just to generate metadata, or does it have any other effect?